### ISSUE TITLE
Many "MenuHelper: menu is null" messages in console.

### ISSUE THREAD

**Issue by jhpedemonte** [maqetta/maqetta#1446]
I'm seeing a lot of "MenuHelper: menu is null" messages in the console, from **MenuHelper.js** line 15.

That line gets called **a lot!**  New Page, drop Dojo button (causes 26 "menu is null" lines); unselect (causes 26 lines); reselect (causes 26 lines).

I'm worried about that code in general.  It subscribes to **"/davinci/ui/widgetSelected"**, which seems to be called often.  26 times is a lot.  Maybe `widgetSelected` is firing too often?

For M4, we should at least remove the **console.log** call.  For M5, there may be some other issues in this code.

**Comment by peller** [comment]
yeah, looks like topic code which is not context-safe. (will only occur if you use a menu widget, of course)  Can you give steps to reproduce this?  A quick test does not turn this up, but I don't doubt that it's getting to this condition.  And what scenario do you have where you get select called 26 times, exactly?

I'm ok with commenting out the log in M4 then kicking this to M5 for a proper fix.

**Comment by peller** [comment]
dup of #1333.  Perhaps you were using the theme editor?  Should we open a separate ticket to investigate select getting called 26 times?  Is that also specific to the theme editor?

**Comment by jhpedemonte** [comment]
I wouldn't say this is a dup of that.  I mainly opened this because of the excessive messages, which I get for **every** widget, not just ones with menus (see 2nd line in description).

At the very least, we should **get rid of the logging messages** -- they get called way too often to be of any use.

Second, we need to investigate that code, because it gets called **26 times** for each individual action.

**Comment by JonFerraiolo** [comment]
@billreed63 

I put a breakpoint at MenuHelper.js:15 and added a couple of menus. What I'm seeing is that each time selection changes, the breakpoint is hit N times where N is the number of menus that ever existed in the session. If I have a new session and create 2 Menu widgets, then the breakpoint gets hit twice. If I save the file and reopen, then click select on a widget, then the breakpoint gets hit 4 times, twice for the old subscriptions and twice for the new subscriptions from the newly (re-)opened file.

To me, this reflects a design problem with our helpers. We have logic now that creates a new helper object instance in various circumstances which allow for N helpers to exist for a particular widget type. Instead, the helpers need to be singleton/static objects where we only load a particular helper once for each session.

**Comment by JonFerraiolo** [comment]
One more comment. Crockford's coding standards say to always put new JS statements on newlines. In this case, we have:

```
if (!menu) { console.log("MenuHelper: menu is null"); return; }
```

by putting everyone on a single line, it is not possible to put a breakpoint on the `console.log` and instead only on the `if`. In this case, not a big problem, but in other cases maybe you only want to have a breakpoint when the conditional is true.

**Comment by jhpedemonte** [comment]
I think we already had a bug open for making it so helpers are only called once.  But I can't find it.

**Comment by peller** [comment]
I don't think that's the problem.  We're talking about the create method getting called on the helper, not the helper instance getting created.  The problem is likely that the corresponding destroy isn't getting called... I think there's an issue for that also somewhere...

**Comment by JonFerraiolo** [comment]
Some discussion in and around this question at #874 and #1151

**Comment by JonFerraiolo** [comment]
@peller, I think the problem here is that we create N instances of the helper class and each instance subscribes to the widgetSelected event. If the helper class became a singleton then we would only have 1 subscription to widgetSelected. (The subscription would need to move out of the Create method and into a constructor or other method that gets called only once per session)

**Comment by peller** [comment]
@JonFerraiolo, I don't think so.  Isn't create() bound to the creation of the widget instance?  There is no correlation between the helper instance and the widget instance (separate issue) which is why I suggested making all the helpers statics, but I don't think that's relevant here.
